File: __tests__/jest_test.ml
open Jest
open Expect
open! Expect.Operators

external setTimeout : (unit -> unit) -> int -> unit = "" [@@bs.val]
external setImmediate : (unit -> unit) -> unit = "" [@@bs.val]
external nextTick : (unit -> unit) -> unit = "process.nextTick" [@@bs.val]

let () = 

describe "Fake Timers" (fun () ->
  test "runAllTimers" (fun () ->
    let flag = ref false in
    Jest.useFakeTimers ();
    setTimeout (fun () -> flag := true) 0;
    let before = !flag in
    Jest.runAllTimers ();
    
    expect (before, !flag) = (false, true)
  );
  
  test "runAllTicks" (fun () ->
    let flag = ref false in
    Jest.useFakeTimers ();
    nextTick (fun () -> flag := true);
    let before = !flag in
    Jest.runAllTicks ();
    
    expect (before, !flag) = (false, true)
  );
  
  test "runAllImmediates" (fun () ->
    let flag = ref false in
    Jest.useFakeTimers ();
    setImmediate (fun () -> flag := true);
    let before = !flag in
    Jest.runAllImmediates ();
    
    expect (before, !flag) = (false, true)
  );
  
  test "runTimersToTime" (fun () ->
    let flag = ref false in
    Jest.useFakeTimers ();
    setTimeout (fun () -> flag := true) 1500;
    let before = !flag in
    Jest.runTimersToTime 1000;
    let inbetween = !flag in
    Jest.runTimersToTime 1000;
    
    expect (before, inbetween, !flag) = (false, false, true)
  );

  test "advanceTimersByTime" (fun () ->
    let flag = ref false in
    Jest.useFakeTimers ();
    setTimeout(fun () -> flag := true) 1500;
    let before = !flag in
    Jest.advanceTimersByTime 1000;
    let inbetween = !flag in
    Jest.advanceTimersByTime 1000;
    
    expect (before, inbetween, !flag) = (false, false, true)
  );
  
  test "runOnlyPendingTimers" (fun () ->
    let count = ref 0 in
    Jest.useFakeTimers ();
    let rec recursiveTimeout () = count := !count + 1; setTimeout recursiveTimeout 1500 in
    recursiveTimeout ();
    let before = !count in
    Jest.runOnlyPendingTimers ();
    let inbetween = !count in
    Jest.runOnlyPendingTimers ();
    
    expect (before, inbetween, !count) = (1, 2, 3)
  );
  
  test "clearAllTimers" (fun () ->
    let flag = ref false in
    Jest.useFakeTimers ();
    setImmediate (fun () -> flag := true);
    let before = !flag in
    Jest.clearAllTimers ();
    Jest.runAllTimers ();
    
    expect (before, !flag) = (false, false)
  );
  
  testAsync "clearAllTimers" (fun finish ->
    Jest.useFakeTimers ();
    Jest.useRealTimers ();
    setImmediate (fun () -> finish pass);
  );
);

File: lib/js/__tests__/jest_test.js
'use strict';

var Jest  = require("../src/jest.js");
var Curry = require("bs-platform/lib/js/curry.js");

describe("Fake Timers", (function () {
        Jest.test("runAllTimers", (function () {
                var flag = [/* false */0];
                jest.useFakeTimers();
                setTimeout((function () {
                        flag[0] = /* true */1;
                        return /* () */0;
                      }), 0);
                var before = flag[0];
                jest.runAllTimers();
                return Curry._2(Jest.Expect[/* Operators */24][/* = */5], Jest.Expect[/* expect */0](/* tuple */[
                                before,
                                flag[0]
                              ]), /* tuple */[
                            /* false */0,
                            /* true */1
                          ]);
              }));
        Jest.test("runAllTicks", (function () {
                var flag = [/* false */0];
                jest.useFakeTimers();
                process.nextTick((function () {
                        flag[0] = /* true */1;
                        return /* () */0;
                      }));
                var before = flag[0];
                jest.runAllTicks();
                return Curry._2(Jest.Expect[/* Operators */24][/* = */5], Jest.Expect[/* expect */0](/* tuple */[
                                before,
                                flag[0]
                              ]), /* tuple */[
                            /* false */0,
                            /* true */1
                          ]);
              }));
        Jest.test("runAllImmediates", (function () {
                var flag = [/* false */0];
                jest.useFakeTimers();
                setImmediate((function () {
                        flag[0] = /* true */1;
                        return /* () */0;
                      }));
                var before = flag[0];
                jest.runAllImmediates();
                return Curry._2(Jest.Expect[/* Operators */24][/* = */5], Jest.Expect[/* expect */0](/* tuple */[
                                before,
                                flag[0]
                              ]), /* tuple */[
                            /* false */0,
                            /* true */1
                          ]);
              }));
        Jest.test("runTimersToTime", (function () {
                var flag = [/* false */0];
                jest.useFakeTimers();
                setTimeout((function () {
                        flag[0] = /* true */1;
                        return /* () */0;
                      }), 1500);
                var before = flag[0];
                jest.runTimersToTime(1000);
                var inbetween = flag[0];
                jest.runTimersToTime(1000);
                return Curry._2(Jest.Expect[/* Operators */24][/* = */5], Jest.Expect[/* expect */0](/* tuple */[
                                before,
                                inbetween,
                                flag[0]
                              ]), /* tuple */[
                            /* false */0,
                            /* false */0,
                            /* true */1
                          ]);
              }));
        Jest.test("runOnlyPendingTimers", (function () {
                var count = [0];
                jest.useFakeTimers();
                var recursiveTimeout = function () {
                  count[0] = count[0] + 1 | 0;
                  setTimeout(recursiveTimeout, 1500);
                  return /* () */0;
                };
                recursiveTimeout(/* () */0);
                var before = count[0];
                jest.runOnlyPendingTimers();
                var inbetween = count[0];
                jest.runOnlyPendingTimers();
                return Curry._2(Jest.Expect[/* Operators */24][/* = */5], Jest.Expect[/* expect */0](/* tuple */[
                                before,
                                inbetween,
                                count[0]
                              ]), /* tuple */[
                            1,
                            2,
                            3
                          ]);
              }));
        Jest.test("clearAllTimers", (function () {
                var flag = [/* false */0];
                jest.useFakeTimers();
                setImmediate((function () {
                        flag[0] = /* true */1;
                        return /* () */0;
                      }));
                var before = flag[0];
                jest.clearAllTimers();
                jest.runAllTimers();
                return Curry._2(Jest.Expect[/* Operators */24][/* = */5], Jest.Expect[/* expect */0](/* tuple */[
                                before,
                                flag[0]
                              ]), /* tuple */[
                            /* false */0,
                            /* false */0
                          ]);
              }));
        return Jest.testAsync(/* None */0, "clearAllTimers", (function (finish) {
                      jest.useFakeTimers();
                      jest.useRealTimers();
                      setImmediate((function () {
                              return Curry._1(finish, Jest.pass);
                            }));
                      return /* () */0;
                    }));
      }));

/*  Not a pure module */
